### PR TITLE
BlobsByRange RPC

### PR DESCRIPTION
**Pull request by kasey**
**What type of PR is this?**
Feature

**What does this PR do? Why is it needed?**
Implements the BlobsByRange RPC handler, specified in the deneb p2p spec: https://github.com/ethereum/consensus-specs/blob/dev/specs/deneb/p2p-interface.md

### CODE REVIEW

File: beacon-chain/sync/metrics.go [hunk 1]
```diff
@@ -82,6 +82,13 @@ var (
 			Buckets: []float64{5, 10, 50, 100, 150, 250, 500, 1000, 2000},
 		},
 	)
+	rpcBlobsByRangeResponseLatency = promauto.NewHistogram(
+		prometheus.HistogramOpts{
+			Name:    "rpc_blobs_by_range_response_latency_milliseconds",
+			Help:    "Captures total time to respond to rpc BlobsByRange requests in a milliseconds distribution",
+			Buckets: []float64{5, 10, 50, 100, 150, 250, 500, 1000, 2000},
```
Comment (terencechain) [inline beacon-chain/sync/metrics.go:89]: We probably want to visit this later but having them the same as block makes sense to me

File: beacon-chain/sync/pending_blocks_queue.go [hunk 2]
```diff
@@ -339,6 +339,19 @@ func (s *Service) sendBatchRootRequest(ctx context.Context, roots [][32]byte, ra
 	return nil
 }
 
+func (_ *Service) dedupRoots(roots [][32]byte) [][32]byte {
```
Comment (terencechain) [inline beacon-chain/sync/pending_blocks_queue.go:342]: interesting we don't have anything like this already implemented

File: beacon-chain/sync/rpc_blob_sidecars_by_root.go [hunk 3]
```diff
@@ -82,6 +82,24 @@ func (s *Service) blobSidecarByRootRPCHandler(ctx context.Context, msg interface
 	return nil
 }
 
+func (s *Service) blobSidecarsByRangeRPCHandler(ctx context.Context, msg interface{}, stream libp2pcore.Stream) error {
```
Comment (terencechain) [inline beacon-chain/sync/rpc_blob_sidecars_by_root.go:85]: note that the file says by root and this is by range, I don't feel strongly though

File: beacon-chain/sync/rpc_blob_sidecars_by_root.go [hunk 4]
```diff
@@ -82,6 +82,24 @@ func (s *Service) blobSidecarByRootRPCHandler(ctx context.Context, msg interface
...
+	ctx, span := trace.StartSpan(ctx, "sync.blobSidecarByRootRPCHandler")
```
Comment (terencechain) [inline beacon-chain/sync/rpc_blob_sidecars_by_root.go:86]: wrong span name

File: beacon-chain/sync/pending_blocks_queue.go (hunk 2 above)
Comment (kasey) [reply beacon-chain/sync/pending_blocks_queue.go:342]: I killed utils.go and re-homed this here because I refactored out everything else in that file.

**Review by terencechain — APPROVED** [review]

File: consensus-types/blocks/roblock.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+	"sort"
+
+	"github.com/pkg/errors"
+	fieldparams "github.com/prysmaticlabs/prysm/v4/config/fieldparams"
+	"github.com/prysmaticlabs/prysm/v4/consensus-types/interfaces"
+)
+
+var ErrRootLength error = errors.New("incorrect length for hash_tree_root")
+
+type ROBlock struct {
```
Comment (rkapka) [inline consensus-types/blocks/roblock.go:13]: I don't like this name. It's very cryptic. Also there is no doc.

**Review by rkapka — CHANGES_REQUESTED** [review]

**Comment by terencechain** [comment]
@rkapka fixed the conflicts, build passed, and the failing tests will be fixed in the subsequent PR

File: consensus-types/blocks/roblock.go (hunk 5 above)
Comment (terencechain) [reply consensus-types/blocks/roblock.go:13]: we can address this change in the `eip4844` branch, and if we ever move this onto the `develop` branch

**Pull request merged**